rpcserver.go

```diff
@@ -1697,6 +1697,28 @@ func validatePayReqExpiry(payReq *zpay32.Invoice) error {
 	return nil
 }
 
+// calculateFeeLimit returns the fee limit in millisatoshis. If a percentage
+// based fee limit has been requested, we'll factor in the ratio provided with
+// the amount of the payment.
+func calculateFeeLimit(feeLimit *lnrpc.FeeLimit,
+	amount lnwire.MilliSatoshi) lnwire.MilliSatoshi {
+
+	switch feeLimit.GetLimit().(type) {
+	case *lnrpc.FeeLimit_Fixed:
+		return lnwire.NewMSatFromSatoshis(
+			btcutil.Amount(feeLimit.GetFixed()),
+		)
+	case *lnrpc.FeeLimit_Percent:
+		return amount * lnwire.MilliSatoshi(feeLimit.GetPercent()/100)
+	default:
+		// If a fee limit was not specified, we'll use the maximum
+		// payment amount as an upper bound since there might be some
+		// routes where you pay more in fees than the amount of the
+		// payment.
+		return maxPaymentMSat
```

Comment:
Maybe we should restrict this to 2x payment amount or something?

Reply:
Yeah I was thinking this too. Not sure if we want to enforce this without adding a `no_fee_limit` flag though.

Reply:
If the caller really wants no fee limit, she can set the fee limit to a ridiculously high number.

Reply:
Still think we need to have a sane default value here.

Reply:
I gave this a spin on testnet and seems o be working as advertised! Only thing I realized is that very small payments might be impossible to get through with this default value, since fees might always be higher. This might be okay though, as you can always increase your fee limit.

Also, for large payments, you might end up paying 2x the amount if the only route is a high fee route (`value + fee` when `fee=value`). This might lead to some unpleasant surprises. Should we set a default, hard coded max value? We could even consider doing this at the `lncli` level. Atm the default value appears to be `0` if you look at `lncli payinvoice -h`. We could change this to a reasonable value?